src/navigation/failure_identification/failure_identification.py

```diff
@@ -0,0 +1,193 @@
+        @param odometry: the current odometry of the rover, has position and velocity data
+
+        publishes a message to the /nav_stuck topic indicating if the rover is stuck
+        """
+
+        TEST_RECOVERY_STATE = get_rosparam("test_recovery_state", False)
+
+        # if the state is 'done' or 'off', write the data frame to a csv file if we were collecting
+        if nav_status.active_states[0] == "DoneState" or nav_status.active_states[0] == "OffState":
+            if self.actively_collecting and self.data_collecting_mode:
+                # print("writing to file")
+                rospy.loginfo("writing to file")
+                if self.path_name is None:
+                    self.write_to_csv()
+                else:
+                    self._df.to_csv(self.path_name, mode="a", header=False)
+                self.actively_collecting = False
+            # return to not collect any data
+            return
+
+        # create a new row for the data frame
+        self.actively_collecting = True
+        cur_row = {}
+        cur_row["row"] = self.row_counter
+        cur_row["time"] = rospy.Time.now()
+
+        # if the stuck button is pressed, the rover is stuck (as indicated by the GUI)
+        if self.cur_stuck:
+            cur_row["stuck"] = 1
+        else:
+            cur_row["stuck"] = 0
+
+        # get the command velocity from the cmd_vel message
+        cur_row["cmd_vel_x"] = self.cur_cmd.linear.x
+        cur_row["cmd_vel_twist"] = self.cur_cmd.angular.z
+
+        # get the x, y, z position of the rover from odometry message
+        cur_row["x"] = odometry.pose.pose.position.x
+        cur_row["y"] = odometry.pose.pose.position.y
+        cur_row["z"] = odometry.pose.pose.position.z
+
+        # get the x, y, z, w rotation of the rover from odometry message
+        cur_row["rot_x"] = odometry.pose.pose.orientation.x
+        cur_row["rot_y"] = odometry.pose.pose.orientation.y
+        cur_row["rot_z"] = odometry.pose.pose.orientation.z
+        cur_row["rot_w"] = odometry.pose.pose.orientation.w
+
+        # get the linear and angular velocity of the rover from odometry message
+        linear_velocity_norm = np.linalg.norm(
+            np.array([odometry.twist.twist.linear.x, odometry.twist.twist.linear.y, odometry.twist.twist.linear.z])
+        )
+        cur_row["linear_velocity"] = linear_velocity_norm  # type: ignore
+        cur_row["angular_velocity"] = odometry.twist.twist.angular.z
+
+        # get the wheel effort and velocity from the drive status message
+        for wheel_num in range(6):
+            cur_row[f"wheel_{wheel_num}_effort"] = drive_status.joint_states.effort[wheel_num]
+            cur_row[f"wheel_{wheel_num}_velocity"] = drive_status.joint_states.velocity[wheel_num]
+
+        # update the data frame with the cur row
+        self._df = pd.concat([self._df, DataFrame([cur_row])])
+        self.row_counter += 1
+
+        if len(self._df) == DATAFRAME_MAX_SIZE:
```

Comment:
shouldn't we only be writing data to the file if we're in collection mode? I imagine during an actual mission we don't need to be writing to csv files

Reply:
You are right, we added this part later and forgot that if statement. This will be fixed.